### PR TITLE
Locking client + volplugin cleanup

### PR DESCRIPTION
**Pull request by erikh**
This isolates the code from volplugin that communicates with the volmaster for clarity and later refactoring purposes.

### CODE REVIEW

**Comment by erikh** [comment]
I've also added porting for the volmaster too.

File: lock/client/client.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+	"github.com/contiv/volplugin/config"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+// Driver is the main force behind the lock module, it controls all methods and
+// several variables.
+type Driver struct {
+	master         string
+	mountStopChans map[string]chan struct{}
```
Comment (mapuri) [inline lock/client/client.go]: I might have missed this, but I am not very clear on advantage of using a `struct{}` vs say `interface{}` or `bool` here. We don't seem to be interpreting/using the `struct` received on this channel anywhere, right?

File: lock/client/client.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	mountMutex     sync.Mutex
```
Comment (mapuri) [inline lock/client/client.go]: minor: just wondering, if we shall just embed the `sync.Mutex` directly here? That way we can just call `d.Lock()` and `d.Unlock()` every, than more verbose `d.mountMutex...`

File: lock/client/client.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+// NewDriver constructs a new driver.
+func NewDriver(master string) *Driver {
+	return &Driver{
+		master:         master,
+		mountStopChans: map[string]chan struct{}{},
```
Comment (mapuri) [inline lock/client/client.go]: minor: I think `make( map[string]chan struct{})` might be more Go-like instantiation of an empty map?

File: lock/lock.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+// NewDriver creates a Driver. Requires a configured TopLevelConfig.
+func NewDriver(config *config.TopLevelConfig, global *config.Global) *Driver {
+	return &Driver{Config: config, Global: global}
+}
+
+// ExecuteWithUseLock executes a function within a lock/context of the passed
+// *config.UseMount.
+func (d *Driver) ExecuteWithUseLock(uc *config.UseMount, runFunc func(d *Driver, uc *config.UseMount) error) error {
```
Comment (mapuri) [inline lock/lock.go]: I think `config.UseMount` in the method args should be replaced with `config.UseLocker`, right?

**Comment by mapuri** [comment]
overall LGTM, just a few minor comments.

File: lock/client/client.go (hunk 1 above)
Comment (erikh) [inline lock/client/client.go]: struct{} is free: http://dave.cheney.net/2014/03/25/the-empty-struct

File: lock/client/client.go (hunk 2 above)
Comment (erikh) [inline lock/client/client.go]: perhaps, but I really want a better reason to rewrite the patch than 
this. :) Maybe in a future refactor.

File: lock/lock.go (hunk 4 above)
Comment (erikh) [inline lock/lock.go]: yeah, you're right.

File: lock/client/client.go (hunk 1 above)
Comment (mapuri) [inline lock/client/client.go]: ah nice, didn't know about this, learned something new :) :+1:

File: lock/client/client.go (hunk 2 above)
Comment (mapuri) [inline lock/client/client.go]: sgtm :)

**Pull request merged**